core/src/main/java/com/digitalpebble/storm/crawler/filtering/basic/QueryParameterNormalizer.java

```diff
@@ -0,0 +1,107 @@
+package com.digitalpebble.storm.crawler.filtering.basic;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import com.digitalpebble.storm.crawler.Metadata;
+import com.digitalpebble.storm.crawler.filtering.URLFilter;
+import com.fasterxml.jackson.databind.JsonNode;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URIBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Basic filter to remove query parameters from urls so parameters that don't change the content of
+ * the page can be removed. An example would be a google analytics query parameter like
+ * "utm_campaign" which might have several different values for a url that points to the same
+ * content.
+ */
+public class QueryParameterNormalizer implements URLFilter {
+
+    private List<String> parametersToRemove;
+    private static final Logger LOG = LoggerFactory.getLogger(QueryParameterNormalizer.class);
+
+    @Override
+    public void configure(Map stormConf, JsonNode filterParams) {
+        JsonNode filenameNode = filterParams.get("parameterNormalizerFile");
+        String rulesFileName = filenameNode.textValue();
+        parametersToRemove = readRules(rulesFileName);
+    }
+
+    @Override
+    public String filter(URL sourceUrl, Metadata sourceMetadata, String urlToFilter) {
+        URIBuilder uriBuilder;
+        try {
+            // Handle illegal characters by making a url first
+            // this will clean illegal characters like |
+            URL url = new URL(urlToFilter);
+            URI uri =
+                    new URI(url.getProtocol(), url.getAuthority(), url.getPath(), url.getQuery(),
+                            url.getRef());
```

Comment:
there is!

if you do toUri() directly, it'll throw a ```URISyntaxException```

but the Uri constructor used above will:
```If a scheme-specific part is given then it is appended. Any character that is not a legal URI character is quoted.```